### DISCUSSION TITLE
Question : Is it possible to count how many logs have been aggregated ?

### DISCUSSION THREAD

**Discussion by nicolasmf** [syslog-ng/syslog-ng#4264] [Q&A] [answered]
Hi,
I would like to aggregate some logs if some fields are the same. In the end I need something like _<log_message> log_count=X_ so I know how many logs have been "fused" into one.

Is there a way to do this ?

**Comment by MrAnno** [comment, marked as answer]
Hi,

You can do that with `grouping-by()`.
The `$(context-length)` template function can help querying the "count" of the aggregation.

Docs:
https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.37/administration-guide/100#TOPIC-1829298

Here is a recent discussion about a similar topic:
https://gitter.im/syslog-ng/syslog-ng?at=638a0935a34b511211599dc7

**Reply by nicolasmf** [reply]
So to fully close this topic, here's my parser config.

```lisp
parser p_aggregate {
        grouping-by(
                key("${.log.devname}/${.log.srcip}/${.log.dstip}/${.log.dstport}")
                aggregate(
                        value("MESSAGE" "${MESSAGE} log_count=$(- $(context-length) 1)")
                )
                timeout(10)
                inject-mode(aggregate-only) # Only send aggregated messages to the output
        );
};```

**Comment by MrAnno** [comment]
@nicolasmf I've converted your issue into a Q&A topic.
We can continue the discussion here.

**Comment by bazsi** [comment]
You may want to use "inherit-mode(aggregate-only)" so that the individual
messages are dropped and only the aggregated ones let through.

**Reply by nicolasmf** [reply]
Hi, I can't find this inherit-mode. What version is this from ?

![image](https://user-images.githubusercontent.com/52258510/210064028-bd6ab45a-d8bd-420d-af92-1207cd77903f.png)

**Reply by bazsi** [reply]
This was introduced in 3.38, unfortunately the documentation is lagging behind at the moment (something I am working on changing)

The first item in the feature list, here: https://github.com/syslog-ng/syslog-ng/releases/tag/syslog-ng-3.38.1

**Reply by nicolasmf** [reply]
Okay, thank you.

It's **_inject-mode_(aggregate-only)** by the way !

**Reply by nicolasmf** [reply]
The problem is when a log is "alone" it doesn't seem to appear in the output. Is there a way to fix that ?

**Reply by bazsi** [reply]
If you specify timeout() for grouping-by(), it's going to emit the message it has accumulated that far at the end of the timeout. Is that solving your issue? I am not sure I completely understand your use-case, I just chimed in as I felt aggregate-only was relevant.

You do have timeout(), so once that elapses, you should see the line that was lingering alone. You can even set timeout to a lower value if 10 seconds is too much for you.

Also, you can request grouping-by() to finish with a state by specifying tigger()

**Reply by nicolasmf** [reply]
First of all, happy new year :) !

So the idea is that :
- I have a lot of log events (10k per second)
- I want to fuse logs that have the same devname srcip dstip dstport by keeping track of how many logs are fused into one with _log_count=X_
- With my current configuration and thanks to your help, I'm able to see if there are more than one log fused into one
- But if a log is "unique" during 10 seconds, I don't see them in the output.

Is there a way to fix that last point ?

Thank you for your time.

**Reply by MrAnno** [reply]
I think this can be achieved by inheriting the last message in the context but only when the context contains 1 message.
I'll try to write a config for this soon.

**Comment by bazsi** [comment]
The unique message should be timing out and reported as log_count=1

Isn't this what you are seeing?

**Reply by nicolasmf** [reply]
Unfortunately, I don't see any _log_count=1_ in my output file, event thought I'm supposed to.

**Reply by nicolasmf** [reply]
After some tests, I noticed that I never see any log_count=1 because the aggregation adds one to the variable for some reason. So I just have to fix it by substracting by one the value of log_count (I'm still looking at how to do it) but it's almost over !

**Comment by bazsi** [comment]
Ah. ok. The new "synthetic" message being generated is added to the context
before aggregation, so you actually have two messages that you aggregate,
the first one is the one which has been collected, the 2nd (last) message
is the one being generated.

This means that you can reference name-value pairs from either, but this
also means that $(context-length) gives you 2.

To subtract one from a number, you can use the "$(- $(context-length) 1)"
template expression.


On Tue, Jan 3, 2023 at 4:51 PM Nicolas MF ***@***.***> wrote:

> After some tests, I noticed that I never see any log_count=1 because the
> aggregation adds one to the variable for some reason. So I just have to fix
> it by decrementing the value of log_count (I'm still looking at how to do
> it) but it's almost over !
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/discussions/4264#discussioncomment-4583141>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5XXL52JW44MOBSFEHLWQRDIPANCNFSM6AAAAAATMC6UDQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Bazsi

**Reply by nicolasmf** [reply]
It works perfectly, thanks, both of you !
